src/browser.ts

```diff
@@ -88,33 +88,109 @@ function hasLegacyDestinations(settings: LegacySettings): boolean {
   )
 }
 
-async function flushBuffered(analytics: Analytics): Promise<void> {
+/**
+ * The names of any Analytics instance methods that can be called pre-initialization.
+ * These methods should exist statically on AnalyticsBrowser.
+ */
+type PreInitMethodName =
+  | 'trackSubmit'
+  | 'trackClick'
+  | 'trackLink'
+  | 'trackForm'
+  | 'pageview'
+  | 'identify'
+  | 'reset'
+  | 'group'
+  | 'track'
+  | 'ready'
+  | 'alias'
+  | 'debug'
+  | 'page'
+  | 'once'
+  | 'off'
+  | 'on'
+  | 'addSourceMiddleware'
+  | 'addIntegrationMiddleware'
+  | 'setAnonymousId'
+  | 'addDestinationMiddleware'
+
+/**
+ *  Represents a buffered method call that occurred before initialization.
+ */
+interface PreInitMethodCall<T extends PreInitMethodName = PreInitMethodName> {
+  method: T
+  args: Parameters<Analytics[T]>
+  called: boolean
+  resolve: (v: ReturnTypeUnwrap<Analytics[T]>) => void
+}
+
+const normalizeSnippetBuffer = (buffer: SnippetBuffer): PreInitMethodCall[] => {
+  return buffer.map(
+    ([methodName, ...args]) =>
+      ({
+        method: methodName,
+        resolve: () => {},
+        args,
+        called: false,
+      } as PreInitMethodCall)
+  )
+}
+
+type SnippetWindowBufferedMethodArgs = any[]
+
+type SnippetWindowBufferedMethodCall = [
+  PreInitMethodName,
+  ...SnippetWindowBufferedMethodArgs
+]
+
+/**
+ * A list of the method calls before initialization for snippet users
+ * For example, [["track", "foo", {bar: 123}], ["page"], ["on", "ready", function(){..}]
+ */
+type SnippetBuffer = SnippetWindowBufferedMethodCall[]
+
+/**
+ * Fetch the buffered method calls from the window object and normalize them.
+ */
+const getSnippetWindowBuffer = (): PreInitMethodCall[] => {
   const wa = window.analytics
   const buffered =
     // @ts-expect-error
-    wa && wa[0] ? [...wa] : []
-
-  for (const [operation, ...args] of buffered) {
-    if (
-      // @ts-expect-error
-      analytics[operation] &&
-      // @ts-expect-error
-      typeof analytics[operation] === 'function'
-    ) {
-      if (operation === 'addSourceMiddleware') {
-        // @ts-expect-error
-        await analytics[operation].call(analytics, ...args)
-      } else {
-        // flush each individual event as its own task, so not to block initial page loads
-        setTimeout(() => {
-          // @ts-expect-error
-          analytics[operation].call(analytics, ...args)
-        }, 0)
-      }
+    (wa && wa[0] ? [...wa] : []) as SnippetBuffer
+  return normalizeSnippetBuffer(buffered)
+}
+
+function flushBuffered(analytics: Analytics, buffer: PreInitMethodCallBuffer) {
+  const callBufferedAnalyticsMethod = (methodCall: PreInitMethodCall) => {
+    const { method } = methodCall
+    // this guard is probably not needed.
+    if (typeof analytics[method] !== 'function') {
+      return console.warn(
+        `invariant error: method call "${method}" does not exist on analytics instance: ${analytics}`
+      )
+    }
```

Comment:
Hmm, I'm also struggling to figure out how this could happen.

Reply:
It was a guard in the existing code! I do think it's an impossible state; if the method doesn't exist, the user will get a reference error.

Reply:
Yeah, probably almost impossible state. Maybe if someone uses a snippet with invalid methods, or manually add :shit: to the buffer.

Reply:
> Yeah, probably almost impossible state. Maybe if someone uses a snippet with invalid methods, or manually add 💩 to the buffer.

Someone would  have to manually update the snippet  copypasta  itself to add a fake method to the string array, and then call that fake method!

Reply:
It's possible that prior to your changes, maybe certain functions wouldn't exist for some edge case, but with all your stricter typings and setup, I think dropping this from future implementations is fine